downloader/artifact_downloader.go

```diff
@@ -133,32 +135,20 @@ func (ad *ConcurrentArtifactDownloader) downloadFile(targetDir, fileName, downlo
 
 	ctx, cancel := context.WithTimeout(context.Background(), ad.Timeout)
 
-	downloader := got.NewWithContext(ctx)
-	downloader.Client = ad.createClient().StandardClient()
-
 	start := time.Now()
 
-	err := downloader.Download(downloadURL, fileFullPath)
-
-	if err != nil && errors.Is(err, context.DeadlineExceeded) {
-		ad.Logger.Warnf("Download duration exceeded %s, second attempt", ad.Timeout)
-
-		cancel()
-
-		start = time.Now()
-
-		ctx, cancel = context.WithTimeout(context.Background(), ad.Timeout)
-		downloader := got.NewWithContext(ctx)
-		downloader.Client = ad.createClient().StandardClient()
-		err = downloader.Download(downloadURL, fileFullPath)
-	}
+	err := downloadWithRetry(ctx, ad.createClient(), downloadURL, fileFullPath, ad.Logger)
 
 	if err != nil {
 		// fallback to single threaded download - the error with the 416 status code seems to happen for 0 size files with got
 		errorMessage := err.Error()
-		if errorMessage == "Response status code is not ok: 416" || strings.Contains(errorMessage, "unexpected EOF") {
+		if strings.Contains(errorMessage, "Response status code is not ok: 416") || strings.Contains(errorMessage, "unexpected EOF") {
 			ad.Logger.Warnf("Multi threaded download failed, switching to single threaded download")
 
+			cancel()
+
+			ctx, cancel = context.WithTimeout(context.Background(), ad.Timeout)
+
```

Comment:
I think having a slower step execution is preferable to a failing step. That is why the single threaded fallback downloader receives a fresh timeout. But with this we could push the execution time up to 10 minutes before eventually arriving to a failed state if the storage provider is flaky.